### ISSUE TITLE
rectangular evaluation lags behind square

### ISSUE THREAD

**Issue by ajtao** [roboflow/rf-detr#715]
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

I'm sharing this in hopes that someone will have an idea for what the problem is ...

I've trained a few rf-detr large models on a custom bbox dataset with approx 20k+ images. These images are all rectangular with aspect ratio H:W of 1:1.7, approx, i.e. like 1080p or 720p. When the training is done, it uses the default image transform settings.

The mAPs are quite strong, compared to other models (yolox) that i've trained, so everything looks fine from that standpoint.

Now comes the issue. I evaluated the trained model with a variety of H,W combinations. The default for large models is for validation to be computed at 704x704. But i varied the H,W across {704, 1024, 1280}.

To my huge suprise, rectangular evaluation dramatically lags behind square evaluation. i.e. it's better to evaluate by squashing these rectangular images into squares than to leave them as native aspect ratio. For example, 1024x1024 (77.0%) resolution eval beats 1024h x 1280w eval (76.3%). That is really weird to me.

| H \ W | 704   | 1024  | 1280  |
|------:|:-----:|:-----:|:-----:|
| 704   | 75.3% | 72.9% | 68.4% |
| 1024  | 72.0% | 77.0% | 76.3% |
| 1280  | 65.9% | 75.9% | 77.2% |

I realize that re-training the model with a stronger bias towards leaving the images with native aspect ratio might change the results, but even so, this result surprised me. It's almost as if there's a bug somewhere.

### Environment

RF-DETR: 1.4.2
Ubuntu: 22.04
Python: 3.12
Pytorch: 2.8.0
GPU: rtx 4090

### Minimal Reproducible Example

My change to coco.py val section looks like this:
I changed coco image transformation such that:

transforms = [T.RandomResize([(args.val_width, args.val_height)], max_size=None)]

### Are you willing to submit a PR?

- [ ] Yes, I'd like to help by submitting a PR!

**Comment by isaacrob** [comment]
yeah rf-detr due to the deformable attention and windowing mechanisms doesn't immediately lend itself to non-square resolutions. there's def some work to be done in this direction but we've held it as out of scope for the time being. so. what you're observing is totally expected and that's why we train with only square images.

**Comment by isaacrob** [comment]
definitely worth noting though that there MAY just be some simple manipulation that makes it work nicely on rectangular images. from a theory perspective there's reason to think it won't immediately transfer so we haven't really looked into it but in practice it might with the right pipelining. if you feel up to it that would be a great thing to contribute, lots of people ask about this

**Comment by ajtao** [comment]
@isaacrob Thanks very much for your reply!

I don't recall seeing such issues with swin-transformer models, which also have windowed attention. Do you happen to know where i could look (papers or code or anything) to understand this issue in rf-detr a bit better? Is this in particular about the positional embedding?

**Comment by isaacrob** [comment]
swin is a different kind of windowing, specifically sliding windowing. the windowing we use here is a much simpler approach that interacts much more natively with accelerators which was introduced in the ViTDet paper and improved upon in LW-DETR. yes, the issue has to do with how the positional encodings interact with windows. but the other major potential issue is the deformable attention. deformable attention creates sampling points normalized from 0 to 1 indicating where they would like to attend in the encoder feature space. see Deformable-DETR for more details. it's not immediately obvious that a deformable attention mechanism would generalize to new input shapes, though it may not be an issue. all this to say we've felt that there's reason to be hesitant that it'll generalize out of the box when trained with only square images so we haven't bothered to look more deeply into it.

the general statement here is that transformer-based models have less inherent inductive bias than other kinds of models such as convolutional models, which can lead to seemingly unintuitive behaviors like this. the flip side is that they work SIGNIFICANTLY better with pretraining and they scale very well. the idea is that at-scale pretraining induces some of the same inductive biases that one might have out of the box with a convolutional network, while also giving them world knowledge. but if, for example, the pretrain only includes square images, it's not obvious that the 'extend to rectangular easily' inductive bias will pop out! for more information on pretraining vs inductive bias I'd recommend reading Hiera (though the speed gains they claim over vanilla ViT appear to disappear when you use optimized runtimes and flash attention, which they explicitly did not include in their benchmarking).

in RF-DETR, we leaned into pretraining to try to make models that can leverage their prior knowledge to transfer extremely well to small and diverse datasets. but that means leaning into these kinds of interesting issues. 

anyway all this to say we kind of assumed it wouldn't work out of the box, which your tests seem to be confirming, but, as with all machine learning research, it's challenging to disambiguate between a confirmation of a theoretical limitation and the existence of a bug or inefficiency in the implementation :) so we definitely welcome any changes that fix this issue while not changing the square image happy path

**Comment by ajtao** [comment]
Very helpful, thank you @isaacrob
